### PR TITLE
proposal for a minimal wraith with . that copes with unicode

### PR DESCRIPTION
**Pull request by ArneBab**
Explanation:

# no single-element special case

The special case of treating a single element on a line as not a
function call leads to the need for more syntax down the line.

Introducing `&` was only required due to relying on this special case.

# . replaces \ and & again

If you use \ at the end of the line, the following is hard to follow:

```
define (display-excitement str reason r2)
  format "I'm SO EXCITED about ~a (~a, ~a)!!"
         string-upcase str \
         reason r2
```

ask your eyes: Is the reason a procedure call?

This becomes clear with the infix `.`:

```
define (display-excitement str reason r2)
  format "I'm SO EXCITED about ~a (~a, ~a)!!"
         string-upcase str
         . reason r2
```

This is the reason why the `.` exists in Wisp. Using the `.` avoids
the need for another syntax-character (the &) interspersed between
every argument.

# in-line indentation and unicode

Matching indentation must always be based on leading whitespace (only
spaces!), because indentation within a line cannot be calculated in
the general case. For example regional indicator symbols (flags) can
be one or two letters wide depending on whether the region is defined
on the system — and may change with system updates.

Therefore brackets cannot have multi-line calls on the first line.

The following could not be parsed in the general case without complex backtracking:

```
define (greeter name)
  let [to-say
          format "Hey there ~a! :D"
                 . name
       show displayfn]
    displayln to-say
```

Therefore to-say must move on its own line so show can be matched up with it:

```
define (greeter name)
  let [
       to-say
          format "Hey there ~a! :D"
                 . name
       show displayfn]
    displayln to-say
```

One way to avoid this would be to state that this would be equivalent
to the code above:

```
define (greeter name)
  let [to-say
       format "Hey there ~a! :D"
                 . name
     show displayfn]
    displayln to-say
```

Upon hitting the less-indented show, it must treat format as nested
into to-say. I did not specify this, because I do not know whether it
can be implemented efficiently. It would mean that

Treating the first element as not-a-function-call would not work,
because a let with multiple arguments could become a valid structure,
even though it is currently an error:

```
(let ((a b c (values 1 2 3))) a)
```

An indentation-sensitive implementation should not prevent the future
use of not-yet-used structures, else it loses the advantage of
s-expressions.

That the let-expression above is currently an error means that the
path is open to giving it specific meaning in a future standard,
because there should not be existing competing implementations.

### CODE REVIEW

File: wraith/wraith.md [hunk 1]
```diff
@@ -171,24 +126,34 @@ for [pet '("cat" "dog" "horse")]          |  (for [(pet '("cat" "dog" "horse"))]
...
-expressions are not permitted to have less indentation than the column
-after the opening parenthesis.
+``` racket
+define (greeter name)                     |  (define (greeter name)
+  let [                                   |    (let (
+       to-say                             |          (to-say
+          format "Hey there ~a! :D"       |           (format "Hey there ~a! :D"
+                 . name]                  |                   name)))
+    displayln to-say                      |      (displayln to-say)))
+```
```
Comment (ArneBab) [inline wraith/wraith.md:153]: this special-case could be avoided with the more complex parser described in the pull-request and comment-description.

File: wraith/wraith.md [hunk 2]
```diff
@@ -171,24 +126,34 @@ for [pet '("cat" "dog" "horse")]          |  (for [(pet '("cat" "dog" "horse"))]
...
+
```
Comment (ArneBab) [inline wraith/wraith.md:154]: this special-case could be avoided with the more complex parser described in the pull-request and comment-description.

**Comment by cwebber** [comment]
I like it, but `&` still does permit something useful with multiple expressions on one line without jumping into `()` land (which doesn't bother me personally).

But I agree this is a more simple/general approach.  I think @AlexKnauth has Opinions (TM).  Would be good to hear them here?

**Comment by AlexKnauth** [comment]
My main problem with using `.` for un-grouping things that would normally be grouped by indentation is that `.` already has a fairly important use in Lisp as the "cons" or "improper list" separator. I really don't like the idea of using `.` with both of those meanings at the same time (trying to make `.` a pun seems like a bad idea), and `.` as "cons" seems more important to me.

However, another character with the same role might work. I think I suggested `@` because it already has a "splicing" connotation, but with at-expressions from Scribble we can't have both.

Having an indentation-splicing or indentation-un-grouping character like `.` here is a good idea, but I don't like using `.` for it.

**Comment by ArneBab** [comment]
I agree that & can be useful, but I think that the added syntax complexity does more damage than good. Since wraith already uses parens for the procedure definition 

```
define (hello who)
  format "Hello ~a!\n" who
```

Having some parens should not worsen the user experience.

But yes, I’d like to hear the opinions of @AlexKnauth on this.

EDIT: I had written this answer but not submitted before I saw yours. Answer to your answer below.

**Comment by ArneBab** [comment]
I’m using `.` not as a pun, but because it is already reserved syntax, though not in that position, and it often already acts as "not a list": 

* In Guile, Kawa, and Emacs Lisp `(. 1)` is just `1`, while 
* In Racket, Chicken, Scheme 48, Chibi, MIT-Scheme, and sbcl `(. 1)` is illegal to use (throws an exception), so there is no conflict.

`(. a b c)` as `a b c` is the generalization of the existing behavior in Guile, Kawa, and Emacs Lisp which does not conflict with allowed use in other Schemes or Common Lisps (I did not test all kinds of lisps, but those are already a large proportion of the ones in use). This also minimizes the chances of conflict with existing libraries.

The larger goal that led to that decision in wisp is to minimize the amount of reserved syntax so macros have as much freedom as possible without introducing ambiguities.

This is also why I like it a lot that you showed that some use of parens and brackets along with consistent indentation parsing inside them removes the need for `:`. If you could get the more complex `greeter` to work in the parser, that would be syntax which is as easy to explain as parenthesis-using s-expressions, keeps all of their power, and is much easier to read for people who aren’t used to Lisps.

**Comment by AlexKnauth** [comment]
A function with a rest argument:
```racket
(define (f a
           b
           . rst)
  rst)
```
I would like this to be able to be written as
```racket
define (f a
          b
          . rst)
  rst
```
Or even as 
```racket
define f a
         b
         . rst
  rst
```
if we keep the "Here the `d` is associated with `b_1` because it has greater indentation than `b_1`" there.

I do want `.` to act as a non-grouping token in the same way keywords do:
```
(f a b
   . rst)
=
(f a b . rst)
and not
(f a b (. rst))
```
but I don't want the `.` to disappear in situations like that.

**Comment by ArneBab** [comment]
In wisp I deal with that by using a double `.`:

```
f a b
   . . rst
=
f a b . rst
=>
(f a b . rst)
=
(f a b
      . rst)
```

Besides: I just checked for the length of the simplest explanations in printed form, and they would fit easily on a single A4-sheet of paper; by folding it and printing two of the following A5 pages per side: [wraith.pdf](https://github.com/cwebber/racket2-rfcs/files/5732200/wraith.pdf)

**Comment by AlexKnauth** [comment]
What about using double-dot as in `..` instead of a single `.` for indentation-splicing?
```racket
define (display-excitement str reason r2)
  format "I'm SO EXCITED about ~a (~a, ~a)!!"
         string-upcase str
         .. reason r2
```

**Comment by ArneBab** [comment]
Double-dot `..` for indentation-splicing sounds good. I don’t see any problems that could arise from it. Since macros typically use triple-dot `...`, and single-dot `.` is forbidden in most places, double-dot is likely to be free. Also it would have saved me some cludges I wrote in [dryads-wake](https://hg.sr.ht/~arnebab/dryads-wake/browse/dryads-wake.w?rev=c677727bbd2c#L1219) to get normal prose working in wisp.

And it’s similar to using `++` for matrix-operations, so it could have the advantage of familiarity.

**Comment by AlexKnauth** [comment]
Okay. I think I'll update the implementation to use `..` double-dot as an indentation-splicing token then. I still think we should keep `&` ampersand for other reasons, but I would be inclined to remove `\` backslash-newline in favor of `..` double-dot, does that sound good to both of you?

**Comment by ArneBab** [comment]
That sounds good to me, yes. Here in the pull-request I added a  self-contained org-mode file from which I compiled the pdf above. Feel free to use that under any license you like.

Depending on your plans with Wraith, you might need something to mark empty indentation levels — that’s where the colon `:` comes from in wisp: It allows using wisp all the way down. But if you don’t mind requiring closing parens for deeper nested double-opening blocks, the paren `(` and bracket `[` syntax with indentation-parsing within would suffice for that.

**Comment by AlexKnauth** [comment]
It seems to me that the `:` is sort of like an open-paren that has its close-paren determined by indentation. Is that right?

I'm not sure I understand what you mean by "empty" indentation levels. However I do prefer having parentheses with indentation-parsing inside them over some non-paren-looking character acting as an open-paren.

**Comment by ArneBab** [comment]
Yes, in wisp the `:` as the only character in a line is an open paren that gets closed when a following line has equal or lower indentation. I think of it as defining an indentation-level: “at this point we have an indentation level, but it has no atom as first element”.

The reason why it exists is to allow for a construct like 

```
let
  :
    foo
      + 1 2
  .. #true
=>
(let
  (
    (foo
      (+ 1 2)))
  #true)
```

Also you can use it for the empty let:

```
let
  :
  .. #true
=>
(let
  ()
  #true)
```
It allows for arbitrary code structures using only indentation. That makes indentation-based wisp a complete replacement for s-expressions: you can use parens, but you do not need them to be able to represent all possible structures.

EDIT: when looking at this code, I wish I had used the double-dot `..` in wisp :) — it has a very nice symmetry with the colon `:`.

**Comment by AlexKnauth** [comment]
I've implemented `..` double-dot as a splicing token, but I haven't removed `\` backslash-newline yet. @cwebber does removing `\` backslash-newline in favor of `..` double-dot sound good to you?

Also there's the question of whether `..` should splice out of parentheses or not.
```
define (display-excitement str reason r2)
  format "I'm SO EXCITED about ~a (~a, ~a)!!"
         string-upcase str
         .. reason r2
```
splicing `.. reason r2` as separate arguments to `format` is fine, but what if it was `(.. reason r2)`? Should it splice the same way parens or not?

**Comment by ArneBab** [comment]
I think it helps to treat indentation-inferred lists and paren-defined lists the same way, because it gives uniform behavior. Also that allows separating the phases: first parsing only indentation and then splicing out lists starting with `..`.

**Comment by ArneBab** [comment]
I think you should splice `(.. reason r2)` just like `.. reason r2`, simply for consistency.

Also I kept thinking: If you add `:`, but restrict its special treatment to „defines an indentation level if it is the only element on the line“, then it does not interfere with its usage in typed racket (since, as far as I understand it, `(:)` has no meaning) and wraith gets the same expressive power as wisp (any tree structure can be represented with indentation). Also that would prevent colon-overload that can look alien to Schemers (it should only be needed in rare cases: only if a very long double indented block is started so you don’t want to track the closing paren).

If I added an option in wisp to use `..` instead of `.`, there would then also be a subset of wraith and wisp that runs in both implementations.

**Comment by AlexKnauth** [comment]
> I think you should splice (.. reason r2) just like .. reason r2, simply for consistency.

I tried to implement that, but I ran into a problem: Once the reader sees an open-paren, the type it expects is a single "tight" result. That's how the types work, and that's also how I discovered my expectations work as a user. As I play around with it in DrRacket I think I have a better idea of why... when I double-click on a paren in DrRacket, it selects the s-expression associated with it, so I can delete, copy-paste, or move it as a single tight unit. That makes grouping by parens feel more powerful than grouping by indentation.

The main reason I don't see `:` as necessary is that `(` or `[` can be used in the same place without changing the indentation structure inside the parens. My main goal for this was to stop making parens a "bad" thing.

**Comment by ArneBab** [comment]
Thank you for trying the `(.. reason r2)` change! I expected Racket to already have support for this, because it is also needed in the case of lists spliced into quasiquoted lists ``` `(,@(list 'reason 'r2))``` or macros ``` #`(#,@'(reason r2))```, but maybe that’s a special-case inside datastructures.

I like the `(` in wraith for shorter blocks. I think it will become problematic for more complex cases, though, because the closing of the block is then again visually far from the opening — getting back to the “how many parens to close”-question (which is one of the problems I see that make it harder to read large-scale paren-structures).

And in general I like wraith a lot. There are some things I would remove (single-argument-lists as "not a call" and "a & a" being equivalent to ".. a a", which are really two sides of the same coin), but I don’t see these as big problems.

There is one thing I think you should really remove, though: In the Drawbacks you state:
> Wraith also does not allow you to write extra parentheses anywhere around any expression without changing meaning.

How is that a Drawback? I think that is actually a strength, because it makes code easier to understand at first glance (no two different constructs to watch out for).

**Comment by ArneBab** [comment]
Datapoints on **single-element-as-not-a-call**: In my wisp-examples, I have 1038 lines that start with a period (`rg -I '^ *\. [^ ]*.*$' | wc -l`). Of these, 202 start with `#:`. 252 contain a single string, 95 use multiple arguments that do not start with #:, and 473 only have one single element on the line (`rg -I '^ *\. [^ ]*.*$' | sed 's/;;.*//g' | rg -v '^ *\. [^ ]+.* * [^ ].*$' -c`).

On the other hand, around 700 lines contain only one element, though many of these are procedures called with the arguments on the next line, so they would not be seen as single elements.

74 of these are actual no-argument procedures, most of them "newline".

So 473 of my lines have a dot-prefix that would not need it with single-argument-as-not-a-call (many of these are the return-value of a procedure), 95 of my lines use multiple arguments prefixed with dot (`.`), 74 are single-argument calls that would need parentheses with single-argument-as-not-a-call. So single-argument-as-not-a-call causes a net-reduction for 400 lines (though visual weight of two parens is much higher than of period-plus-space).

The total number of lines I checked is almost 10.000.

So there is a trade-off: Having single-element lines as values (not as calls) reduces line-noise for 4% of the lines, but it adds inconsistency. The problems in readable (not being able to continue the argument list without ugly constructs) are avoided in wraith by the `..` construct.

And it is an actual trade-off: as I see it from looking at the data (and my experience), the answer is down to personal taste. I value consistency more, because that’s what brought me to Scheme in the first place.

**Comment by ArneBab** [comment]
I took this chance to go back to the mental design drawing board of wisp. I won’t do those changes there, because I believe that something that’s been out in the wild for years should not do breaking changes, but I do see things that could make a more polished language (which could be your wraith).

Having parens break out of indentation handling is a win for re-using code with parentheses, but it is a loss for uniformity (along with one of the top feature requests I got and declined — process indentation in parens), and it requires changing multiple statements when switching from non-paren to paren. It did help adoption (“you can just re-use any scheme in parens by prefixing it with `. `” (I made lots of use of that), but the language would be cleaner without it.

On the other hand, after going back to it, I no longer see `..` as clearly better that `.`, because it breaks alignment compared with `:`. Though that would be less of an issue with wisp-processing inside parens (because that would reduce the use of `:`).

Finally, the choice to implement `{a + b}` as non-inlining (as `(+ a b)` that turns into `((+ a b))` when alone on its line) makes wisp-code less uniform than it would be otherwise (especially in cond — it looks neat in there, but it is different from the surrounding code). This is the only language-level argument I see in favor of treating single-element-lines specially.

Thank you for getting me there!

I hope wraith will still get a bit simpler (as I wrote in previous comments) — and I hope it will then also get into Guile!

**Comment by cwebber** [comment]
We could always use the ellipses character: `…` (U+2026) ;)

I'm joking though, I think using ascii characters is probably better.  Though, arguably, a wraith-mode could turn `...` automatically into `…`.  Still probably not a good idea, though.

I like `..` or `.` and I'm in favor of removing `\`; I think it leads to unexpected problems.

**Comment by ArneBab** [comment]
(sidenote: I’m now linking to wraith from https://www.draketo.de/software/wisp#similar-projects — I hope that’s OK; if not, I’ll remove that link again )

**Comment by ArneBab** [comment]
Do you have updates on wraith?

**Comment by AlexKnauth** [comment]
Not since I updated the documentation for `..` double-dot

**Comment by ArneBab** [comment]
Is this then fully usable already? Can I fire up racket and play with wraith?

**Comment by ArneBab** [comment]
Any news?

**Comment by AlexKnauth** [comment]
The [`wraith` branch `wraith` directory](https://github.com/cwebber/racket2-rfcs/tree/wraith/wraith) is usable as a racket package that defines `#lang wraith` as an alternative reader that takes a module-language as an argument, such as `#lang wraith racket`.

It should be installable with `raco pkg install git://github.com/cwebber/racket2-rfcs?path=wraith#wraith`.

The ideas in this PR include (1) the splicing `.`, and (2) making indentation levels of tokens on the same line after non-whitespace tokens no longer relevant. We've talked about the splicing `.` and changing it to `..`, which is how it is implemented now in the `wraith` branch.

I don't think we've talked much about (2), except to say that it's a problem that unicode characters that take up more space than 1 "monospace character" cause problems for alignment. I think we had 3 main approaches to solving this:
 - (a) Make those indentation levels no longer relevant, as this PR does
 - (b) Assume everyone is using a truly monospace font, and keep those indentation levels even if it looks misaligned in a duospace font. (blame the user for using a duospace font instead of a monospace font)
 - (c) Support duospace fonts in port-column-counters, editor-column-counters, and editor-autoindenters.

My preference is obviously (c), but (b) is what's currently implemented.

**Comment by ArneBab** [comment]
Thank you for your answer!

The problem with (b) and (c) is that you cannot ever know whether a given element will parse as singe or double element. It’s broken in the unicode spec. It might change depending on whether the country whose laws apply to your device recognizes a given country or not.

Did you already present #lang wraith to a larger audience somewhere? If not: Is there something you want to do before you do so?

**Comment by AlexKnauth** [comment]
Well I had planned to try to use the character-widths in the GNU Unifont to determine whether to increment column numbers by 1 or 2 on a character, but I'm not sure how to do that and I didn't get very far last time I tried to dig into the Racket Ports' line-and-column-counting systems.

I got this function at least:
```
;; string-width : String -> Natural
(define (string-width s)
  (define-values [w h b v]
    (send (new bitmap-dc% [bitmap (make-object bitmap% 1 1)])
          get-text-extent
          s
          (make-object font% 2 'modern)))
  (exact-ceiling w))
```

**Comment by AlexKnauth** [comment]
I also tried
```
(define-ffi-definer define-ffi (ffi-lib "libffi"))
(define-ffi wcwidth (_fun _wchar -> _int))
```

but that didn't seem to work properly, the function kept returning `-1` for a lot of unicode characters that I would expect answers of `1` or `2` for, at least on my machine.

**Comment by cwebber** [comment]
[This post on StackOverflow](https://stackoverflow.com/questions/3634627/how-to-know-the-preferred-display-width-in-columns-of-unicode-characters) looks like it could be helpful?

**Comment by cwebber** [comment]
BTW I've been using Wraith with slideshow.  It's a very good experience.

Even with this PR open, I wonder if it's time to talk about Wraith to a racket-users type audience?  Get people interested and talking about it...

**Comment by martin12333** [comment]
just a backlink: I have added to reddit r/whitespaceLisp a link to this PR (as a link to Wraith)

https://www.reddit.com/r/whitespaceLisp/comments/n4smbx/proposal_for_a_minimal_wraith_with_that_copes/

**Comment by ShalokShalom** [comment]
I am curious, if this project is still succeeded anywhere, and if somebody knows an alternative, in case that is not the case.
